src/components/structures/MessagePanel.js

```diff
@@ -471,13 +471,19 @@ module.exports = React.createClass({
         return ret;
     },
 
-    _wantsDateSeparator: function(prevEvent, nextEventTs) {
+    _wantsDateSeparator: function(prevEvent, nextEvent) {
         if (prevEvent == null) {
             // first event in the panel: depends if we could back-paginate from
             // here.
             return !this.props.suppressFirstDateSeparator;
         }
-        return Math.floor(prevEvent.getTs() / MILLIS_IN_DAY) !== Math.floor(nextEventTs / MILLIS_IN_DAY);
+        // Return early for events that are > 24h apart
+        if (Math.abs(prevEvent.getTs() - nextEvent.getTs()) > MILLIS_IN_DAY) {
+            return true;
+        }
+
+        // Compare weekdays
+        return prevEvent.getDate().getDay() !== nextEvent.getDate().getDay();
```

Comment:
Why does it feel odd? After all, the thing actually displayed is the day of the week. There's no reason other than it happens to be one way of doing it that works. Also, comparing the date would require getting day, month, year and comparing them all or generating a locale string.